bdai_ros2_wrappers/test/test_service_handle.py

```diff
@@ -0,0 +1,250 @@
+            self._keep_spinning = True
+            try:
+                while self._context.ok() and self._keep_spinning:
+                    self._executor.spin_once(timeout_sec=0.1)
+            except (ExternalShutdownException, KeyboardInterrupt):
+                pass
+            self._keep_spinning = False
+
+    def stop(self) -> None:
+        self._keep_spinning = False
+        if self._executor is not None:
+            self._executor.shutdown()
+            self._executor.remove_node(self)
+            self._executor = None
+        if self._thread is not None:
+            self._thread.join()
+            self._thread = None
+        self._service = None
+
+
+class ServiceHandleTriggerTest(unittest.TestCase):
+    def setUp(self) -> None:
+        rclpy.init()
+        self.client_node: Optional[Node] = Node("TriggerClient")
+        self.service_name = "trigger_service"
+        self.service_type = Trigger
+        self.client = self.client_node.create_client(
+            self.service_type, self.service_name, callback_group=ReentrantCallbackGroup()
+        )
+        self.server_node: Optional[FooBarService] = None
+
+    def tearDown(self) -> None:
+        if self.client_node is not None:
+            self.client_node.destroy_node()
+            self.client_node = None
+        if self.server_node is not None:
+            self.server_node.stop()
+            self.server_node.destroy_node()
+        self.server_node = None
+        rclpy.shutdown()
+
+    def _internal(self, name: str) -> Tuple[SrvTypeResponse, bool]:
+        self.assertTrue(self.client.wait_for_service(1))
+        failure = False
+
+        def _failure_callback() -> None:
+            nonlocal failure
+            failure = True
+
+        handle = ServiceHandle(name)
+        handle.set_on_failure_callback(_failure_callback)
+
+        req = self.service_type.Request()
+        service_future = self.client.call_async(req)
+        handle.set_send_service_future(service_future)
+
+        while handle.result is None:
+            rclpy.spin_once(
+                self.client_node,
+            )
+        return handle.result, failure
+
+    def test_success(self) -> None:
+        """Tests if the service call is a success"""
+        self.server_node = FooBarService(self.service_type, self.service_name, callback=_callback_trigger_success)
+        result, failure = self._internal("trigger_success")
+        self.assertTrue(result.success)
+        self.assertFalse(failure)
+        self.assertEqual(result.message, "foo")
+
+    def test_failure(self) -> None:
+        """Tests if the service call is a failure"""
+        self.server_node = FooBarService(self.service_type, self.service_name, callback=_callback_trigger_fail)
+        result, failure = self._internal("trigger_failure")
+        self.assertFalse(result.success)
+        self.assertTrue(failure)
+
+
+class ServiceHandleSetBoolTest(unittest.TestCase):
+    def setUp(self) -> None:
```

Comment:
I see a lot of *similar* `setUp`, `tearDown`, and other functions.  (Not sure how similar or if they are identical).  Is there a way to pull this stuff into something common to avoid the duplication of code?

Reply:
I have tried to make them all inherit from a single test case class but I was having some trouble. I can give it a try again and see if there's a specifi way to do it.

Reply:
I am more familiar with `TEST_F` in gtest.  Not sure if there is something similar in pytest...

Reply:
Was able to get around it using some sneaky inheritance hacks